### PR TITLE
Reduce TSV loader complexity

### PR DESCRIPTION
**Pull request by AetherUnbound**
## Description

This PR attempts to reduce the DAG complexity of the TSV loader by removing redundant steps & consolidating code.

TL;DR: Extra steps were present which were covered by Airflow functionality and have now been removed.

### Technical explanation

There were a number of steps in this DAG that were `DummyOperator`s (no-op steps). From what I could tell, the only purpose of these operators was to use Airflow's Trigger Rules[^1]. Incidentally, the same rules were also added to the actual operational tasks within the DAG, which ended up being a bit redundant. I've removed all of these no-op steps, and I've tried to make the logic a bit clearer (using the `doc_md` attribute for each task). The essence of these changes is:

* The table cleanup task will always run after the create table step (regardless of other failures) to ensure the temporary table gets cleaned up.
* If any of the critical pieces of the loader DAG fail, the TSV will be moved to a "failed" location to be dealt with by an operator manually.
* The TSV will be deleted locally **only after** the data has been successfully loaded into S3. This will preserve the TSV under all failure circumstances, since it is the most expensive step to redo.

There was also a `load_local_data` step present in the original DAG. From the looks of it, this step was intended to run if there were any failures uploading the file to S3; it would attempt to load the data directly from the local machine into Postgres, circumventing S3. I removed this for two reasons:
1. We absolutely want to preserve the data in S3 where possible. I would rather have the data exist in S3 and be loaded into the database at a later time than have the data loaded into the database and no record of it in S3.
1. The specific set of trigger rules applied to this step and its parents meant that this step **would never be run**, it would always be skipped :sweat_smile:. 

I restructured the DAG code a bit to make the pieces easier to grok. There's no need to have separate functions in the `operators` file with the sole purpose of generating a `PythonOperator` or similar - that code can be defined directly in the DAG file. It can also be difficult to tell what the arguments are for `PythonOperator` steps, since we don't get to see the function signature in the DAG file. In order to help with this, I've made all the arguments to the `python_callable` explicit via `op_kwargs` (rather than `op_args`). I discovered that there's a `doc_md` field on tasks as well as DAGs (visible from the task's `Instance Details` page), which serves as a handy place for documentation too.

The first TSV staging step was set up to always succeed and pass along any files (if any) that it found. The results were then used to determine whether the rest of the DAG should be run, which is part of the `ShortCircuitOperator`. This logic itself can be shortcut by raising an `AirflowSkipException` within the called function. This also gives us a clearer indication of what's happening with a given run: if no files were found everything gets skipped since there's nothing to do. Having the first task succeed while all others are skipped seems to indicate that there was some part of the loader that ran while the rest was unneeded, which felt a bit confusing to me.

Lastly, I changed the `max_active_runs` on this DAG to `1`. The DAG is set up to move files out of the main output directory as soon as it identifies that they can be worked on, but there's still a potential race condition where more than one DAG would attempt to process the same file. This will be addressed in #285, in which case we can bump the number of simultaneous runs back up.

Whew, lots of words, thanks if you made it this far :smile:

[^1]: https://airflow.apache.org/docs/apache-airflow/1.10.12/concepts.html#trigger-rules

### Screenshots

**Before**
image

**After**
Screenshot_2021-11-08_13-50-55

## Testing Instructions

1. Run a provider workflow long enough to have it write to disk, then mark it as successful.
1. Enable the `tsv_to_postgres_loader` DAG and wait for it to pick up the TSV generated in step 1>
1. Watch as the DAG (hopefully) succeeds!

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [ ] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by zackkrida — APPROVED** [review]
Works great. Ran a few DAGs and confirmed everything worled as expected:
CleanShot 2021-11-08 at 18 28 59@2x

Then `docker stop`-ed the `minio` container, ran another DAG, and confirmed that the failures are moved correctly:

CleanShot 2021-11-08 at 18 40 34@2x

```shell
# note, my local `.env` still had `/tmp` for the output dir
airflow@405c1da0932c:/tmp/db_loader_failures/20211108T233800$ ls
wordpress_image_v001_20211108233707.tsv
```

**Review by krysal — APPROVED** [review]
So nice to have this DAG simplified _and_ documented ✨ It worked perfectly.

And having the Minio UI to inspect the bucket is quite handy too 👍

CleanShot 2021-11-10 at 14 01 12@2x
